jumanji/environments/packing/jigsaw/env.py

```diff
@@ -0,0 +1,527 @@
+
+        return legal
+
+    def _get_ones_like_expanded_piece(self, grid_piece: chex.Array) -> chex.Array:
+        """Makes a grid of zeroes with ones where the piece is placed.
+
+        Args:
+            grid_piece: piece placed on a grid of zeroes.
+        """
+
+        grid_with_ones = jnp.where(grid_piece != 0, 1, 0)
+
+        return grid_with_ones
+
+    def _expand_piece_to_board(
+        self,
+        piece: chex.Array,
+        row_coord: chex.Numeric,
+        col_coord: chex.Numeric,
+    ) -> chex.Array:
+        """Takes a piece and places it on a grid of zeroes with the same size as the board.
+
+        Args:
+            state: current state of the environment.
+            piece: piece to place on the board.
+            row_coord: row coordinate on the board where the top left corner
+                of the piece will be placed.
+            col_coord: column coordinate on the board where the top left corner
+                of the piece will be placed.
+
+        Returns:
+            Grid of zeroes with values where the piece is placed.
+        """
+
+        grid_with_piece = jnp.zeros((self.num_rows, self.num_cols), dtype=jnp.float32)
+        place_location = (row_coord, col_coord)
+
+        grid_with_piece = jax.lax.dynamic_update_slice(
+            grid_with_piece, piece, place_location
+        )
+
+        return grid_with_piece
+
+    def _observation_from_state(self, state: State) -> Observation:
+        """Creates an observation from a state.
+
+        Args:
+            state: State to create an observation from.
+
+        Returns:
+            An observation.
+        """
+
+        return Observation(
+            current_board=state.current_board,
+            action_mask=state.action_mask,
+            pieces=state.pieces,
+        )
+
+    def _expand_all_pieces_to_boards(
+        self,
+        pieces: chex.Array,
+        piece_idxs: chex.Array,
+        rotations: chex.Array,
+        rows: chex.Array,
+        cols: chex.Array,
+    ) -> chex.Array:
+        """Takes multiple pieces and their corresponding rotations and positions,
+            and generates a grid for each piece.
+
+        Args:
+            pieces: array of possible pieces.
+            piece_idxs: array of indices of the pieces to place.
+            rotations: array of all possible rotations for each piece.
+            rows: array of row coordinates.
+            cols: array of column coordinates.
+        """
+
+        batch_expand_piece_to_board = jax.vmap(
+            self._expand_piece_to_board, in_axes=(0, 0, 0)
+        )
+
+        all_possible_pieces = pieces[piece_idxs]
+        rotated_pieces = jax.vmap(rotate_piece, in_axes=(0, 0))(
+            all_possible_pieces, rotations
+        )
+        grids = batch_expand_piece_to_board(rotated_pieces, rows, cols)
+
+        batch_get_ones_like_expanded_piece = jax.vmap(
+            self._get_ones_like_expanded_piece, in_axes=(0)
```

Comment:
If you delete `_get_ones_like_expanded_piece` then:

```suggestion
            lambda x: x != 0, in_axes=(0)
```